rhel/vex/fetcher.go

```diff
@@ -0,0 +1,288 @@
+		} else {
+			if err := f.Close(); err != nil {
+				zlog.Warn(ctx).Err(err).Msg("unable to close spool")
+			}
+		}
+	}()
+
+	if fp.changesEtag == "" { // Used to inform whether this is the first run.
+		// We need to go after the full corpus of vulnerabilities
+		// First we target the archive_latest.txt file
+		latestURI, err := u.url.Parse(latestFile)
+		if err != nil {
+			return nil, hint, err
+		}
+		latestReq, err := http.NewRequestWithContext(ctx, http.MethodGet, latestURI.String(), nil)
+		if err != nil {
+			return nil, hint, err
+		}
+		latestRes, err := u.client.Do(latestReq)
+		if latestRes != nil {
+			defer latestRes.Body.Close()
+		}
+		if err != nil {
+			return nil, hint, err
+		}
+		if latestRes.StatusCode != http.StatusOK {
+			return nil, hint, fmt.Errorf("unexpected response from archive_latest.txt: %s", latestRes.Status)
+		}
+
+		body, err := io.ReadAll(latestRes.Body) // Fine to use as expecting small number of bytes.
+		if err != nil {
+			return nil, hint, err
+		}
+
+		compressedFilename := string(body)
+		zlog.Debug(ctx).
+			Str("filename", compressedFilename).
+			Msg("requesting latest compressed file")
+
+		// Decided against parsing the filename to allow the implementation
+		// to change under the hood w/o Clair caring.
+		uri, err := u.url.Parse(compressedFilename)
+		if err != nil {
+			return nil, hint, err
+		}
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, uri.String(), nil)
+		if err != nil {
+			return nil, hint, err
+		}
+
+		u.client.Timeout = time.Minute * 2 // This is a large request.
+		res, err := u.client.Do(req)
+		if res != nil {
+			defer res.Body.Close()
+		}
+		if err != nil {
+			return nil, hint, err
+		}
+		if res.StatusCode != http.StatusOK {
+			return nil, hint, fmt.Errorf("unexpected response from latest compressed file: %s", res.Status)
+		}
+
+		lm := res.Header.Get("last-modified")
+		fp.requestTime, err = time.Parse(http.TimeFormat, lm)
+		if err != nil {
+			return nil, hint, fmt.Errorf("could not parse last-modified header %s:, %w", lm, err)
+		}
+		z, err := zstd.NewReader(res.Body)
+		if err != nil {
+			return nil, hint, err
+		}
+		defer z.Close()
+		r := tar.NewReader(z)
+
+		var (
+			h              *tar.Header
+			buf, bc        bytes.Buffer
+			entriesWritten int
+		)
+		for h, err = r.Next(); err == nil; h, err = r.Next() {
+			if h.Typeflag != tar.TypeReg {
+				continue
+			}
+			year, err := strconv.ParseInt(path.Dir(h.Name), 10, 64)
+			if err != nil {
+				return nil, hint, fmt.Errorf("error parsing year %w", err)
+			}
+			if year < lookBackToYear {
+				continue
+			}
+			buf.Grow(int(h.Size))
+			if _, err := buf.ReadFrom(r); err != nil {
+				return nil, hint, err
+			}
+
+			err = json.Compact(&bc, buf.Bytes())
+			if err != nil {
+				return nil, hint, fmt.Errorf("error compressing JSON %s: %w", h.Name, err)
+			}
+			bc.WriteByte('\n')
+			cw.Write(bc.Bytes())
+			buf.Reset()
+			bc.Reset()
+			entriesWritten++
+		}
+		if err != io.EOF {
```

Comment:
similar to another question I had: I've seen a lot of `errors.Is(err, io.EOF)` in the codebase